### DISCUSSION TITLE
How to write a script that starts Toggl timer, sets Slack status and sleeps display?

### DISCUSSION THREAD

**Discussion by jsjoeio** [johnlindquist/kit#49] [Q&A] [answered]
I'm new to ScriptKit and thought it might help with a task that I do everyday: eat lunch.

When I eat lunch, I always follow these three steps:
1. Start [Toggl Track timer](https://toggl.com/track/) with "lunch"
2. Set Slack status to "🍔 fueling up" 
3. Sleep display 

I think ScriptKit could help with this. How difficult would it be do to this?

## More info

- looks like Toggl has an API for [creating a time entry](https://github.com/toggl/toggl_api_docs/blob/master/chapters/time_entries.md#create-a-time-entry)

_Note: question was originally asked as a Twitter DM. Posting here in case it helps others!_

**Comment by johnlindquist** [comment, marked as answer]
## Toggl Script

I don't know _exactly_ what you want to create with the Toggl API, but here's an example that create a "time entry".

```js
let { format } = await npm("date-fns")
let btoa = await npm("btoa")
let TOGGL_TOKEN = await env("TOGGL_TOKEN") //Go here -> https://track.toggl.com/profile
let wid = await env("WORKSPACE_ID") //The number in your URL, e.g. 5198420 -> https://track.toggl.com/projects/5198420/list


let start = format(new Date(), "yyyy-MM-dd'T'HH:mm:ssxxx")

let duration = (await arg("How many minutes")) * 60

let response = await post(
  "https://api.track.toggl.com/api/v8/time_entries",
  {
    time_entry: {
      description: "Lunch",
      created_with: "Script Kit",
      start,
      duration,
      wid,
    },
  },
  {
    headers: {
      "Content-Type": "application/json",
      Authorization:
        "Basic " + btoa(`${TOGGL_TOKEN}:api_token`),
    },
  }
)

```

## Slack Script
```js
let Slack = await npm("slack")
//https://api.slack.com/apps 
// Create an app. 
// In the menu -> OAuth & Permssions. 
// User Token Scopes. Add "users.profile:write"
// Grab the "User OAuth Token"
let slack = new Slack({ token: await env("SLACK_TOKEN") })

let profile = {
  status_text: "Out to lunch",
  status_emoji: ":hamburger:",
}

slack.users.profile.set({
  profile,
})
```

## Sleep Script
```js
let { lock } = await kit("system")

await lock()
```

## Go to Lunch Script

This script runs the other 3 scripts in a single command:

```js
//Menu: Go to Lunch
//Description: Create toggl, set slack, sleep computer

await run("toggl-lunch")
await run("set-slack-lunch")
await run("lunch-sleep")
```

**Reply by jsjoeio** [reply]
This is so readable! Thanks so much for the help @johnlindquist ! 🙌

**Reply by jsjoeio** [reply]
Hmm..I'm not sure what I did wrong, but I can't get the Sleep Script to work. Seems like it couldn't find the `system` module? 

![2021-03-11 20 10 50](https://user-images.githubusercontent.com/3806031/110886486-f2d21100-82a5-11eb-9d51-7978ee43efa8.gif)

Are there logs written somewhere to look what went wrong?

**Reply by jsjoeio** [reply]
The Slack Script worked though! 🎉

<img width="194" alt="image" src="https://user-images.githubusercontent.com/3806031/110887016-c66ac480-82a6-11eb-9604-d861e75c0dd9.png">

**Reply by johnlindquist** [reply]
logs are in the `~/.kenv/logs` dir.  (You can see the dir in the explorer at the bottom of your screenshot.

The `console.log` will show you anything you've console.log'd in a script
The `kit.log` will show if anything happened in the app.

I forgot to commit the `system.js` (which includes the `lock` func). It's committing now and I started another build.

You can `update` by right-clicking the icon (or `cmd + shift + ;`) and typing "update".
Note: if you're on a build earlier than `Kit-1.1.0-beta.63.dmg` (today's build), you may have to click the icon after you update. Then quit again for it to "really" update. This should all be fixed in 1.1.0-beta.63+

**Reply by jsjoeio** [reply]
Ah! Didn't even see that 🤦‍♂️

Awesome! Good to know about both `console.log` and `kit.log`. It works! 

And the update worked just fine. And now `lock` works as expected 🚀

**Reply by jsjoeio** [reply]
I also opened a PR adding a `sleep` utility function to `kit/system.js`: https://github.com/johnlindquist/kit/pull/50

**Reply by jsjoeio** [reply]
I ended up using a different endpoint from the Toggl API:

## Toggl Script
```javascript
let btoa = await npm("btoa")
let TOGGL_TOKEN = await env("TOGGL_TOKEN") //Go here -> https://track.toggl.com/profile
let wid = await env("WORKSPACE_ID") //The number in your URL, e.g. 5198420 -> https://track.toggl.com/projects/5198420/list

let response = await post(
  "https://api.track.toggl.com/api/v8/time_entries/start",
  {
    time_entry: {
      description: "lunch",
      created_with: "Script Kit",
      // This is the project id for "Coder"
      // e.g. 165657166 -> https://track.toggl.com/2151494/projects/165657166/team
      pid: "165657166",
      wid,
    },
  },
  {
    headers: {
      "Content-Type": "application/json",
      Authorization:
        "Basic " + btoa(`${TOGGL_TOKEN}:api_token`),
    },
  }
)
```
